ci-automation/tapfile_helper_lib.sh

```diff
@@ -197,27 +197,167 @@ function tap_failed_tests_for_vendor() {
 }
 # --
 
+# TAP output format primitives for tap_generate_report()
+
+__tap_print_header() {
+    local arch="$1"
+    local version="$2"
+    local vendors="$3"
+    local count="$4"
+
+    # We use count + 1 here because the very first "test result" will just print
+    #  the list of platforms tested, not an actual test's result.
+    echo "1..$((count+1))"
+    echo "ok - Version: ${version}, Architecture: ${arch}"
+    echo "   ---"
+    echo "   Platforms tested: ${vendors}"
+    echo "   ..."
+}
+# --
+
+__tap_print_test_verdict() {
+    local verdict="$1"
+    local name="$2"
+    local succeded_vendors="$3"
+    local failed_vendors="$4"
+
+    echo "${verdict} - ${test_name}"
+    echo "   ---"
+
+    if [ -n "${succeded_vendors}" ] ; then
+        echo "   Succeeded: ${succeded_vendors}"
+    fi
+    if [ -n "${failed_vendors}" ] ; then
+        echo "   Failed: ${failed_vendors}"
+    fi
+}
+# --
+
+__tap_print_test_run_diag_output() {
+    local vendor="$1"
+    local run="$2"
+    echo "   Error messages for ${vendor}, run ${run}:"
+    cat -
+}
+# --
+
+__tap_finish_test_verdict() {
+    local verdict="$1"
+    local name="$2"
+    local succeded_vendors="$3"
+    local failed_vendors="$4"
+    echo "   ..."
+}
+# --
+
+__tap_finish_test_report() {
+    true
+}
```

Comment:
`finish_test_report` vfunc seems to be unnecessary, as both `tap` and `md` just return `true`. Are there plans for some other format, where `finish_test_report` would be useful?

Reply:
Well, I guess it will be useful when implementing json format. Nevermind.

Reply:
Correct! @gabriel-samfira mentioned a desire for JSON at some point.